### ISSUE TITLE
Switch to Plugin API, convert gesture detection to a layer

### ISSUE THREAD

**Issue by johnpryan** [fleaflet/flutter_map#174]
1. making all layers a "plugin" and 2. adding a gesture detection layer.

see #160 for background

**Comment by kengu** [comment]
@johnpryan: I'm currently blocked on this issue. Could we start an design discussion so that I can attempt an implementation of this issue?

**Comment by kengu** [comment]
From #153 

> After looking at this a little more closely, I don't think this PR should make changes to MapGestureMixin. It should be each layer's job to enter the gesture arena and immediately leave the gesture arena if the tap (or drag, for example) gesture is not on a polygon. This will be more flexible in the long run and this approach will allow for more interactive gestures by simply leaving it up to the layers themselves. 
>
> There are some challenges to doing this; the main challenge is that FlutterMapState is currently handling the gestures, but really the gesture mixing should be just a layer rendered above the tile layer.
>
>I am going to make a separate ticket for this but for now I would like to leave this PR open so that we can come back and add this feature after 1. making all layers a "plugin" and 2. adding a gesture detection layer.

**Comment by kengu** [comment]
@johnpryan, does "entering the gesture area" imply that each layer should implement a `RawGestureDetector` along the same lines as [Flutter Deep Dive: Gestures](https://medium.com/flutter-community/flutter-deep-dive-gestures-c16203b3434f). Or is it just an analogy, implying something more along the lines of each layer listening for gesture events generated by the gesture layer and handling them? 

If it is the latter, I guess this could easily be implemented with streams passed to each layer with the `createLayer` in the `MapState`object. Each layer could then listen for specific events using the `filter` method, and convert hits on geometries to invocation of callbacks passed to the layer in `LayerOptions`. This would also solve my current problem, which is implementing a geometry editor plugin.

**Comment by johnpryan** [comment]
Yes, each layer should have the ability to handle gestures for themselves without relying on the MapGestureMixin to decide what polygon, circle, or marker was tapped. The goal is to support tap events of all layer types. https://github.com/johnpryan/flutter_map/pull/160 does not quite work for cases like that

**Comment by ruizalexandre** [comment]
@johnpryan I pushed a PR to move layers to Flutter widgets ! With that you can rebuild layer easily ! Solve performance issue also 

#542 here
